design/003-service-topology-support.md

```diff
@@ -0,0 +1,118 @@
+    - Every address in the address set should now retain topology metadata. Since the order is not specific, it can be incorporated in the labels map that is already part of the struct.
+
+* **Service topology**: extend the current `servicePublisher` implementation to include the topology preference as described in the service spec. This implies making changes to the handler functions of the watcher.
+
+
+* **Adding weights to addresses**: to add weights to an address, we need access to the service topology, the source node (its topology labels) and the endpoint topology. If the source node name is sent by the proxy, the node's topology values can be trivially accessed using the `K8s API`. There are two edge cases that exist under my assumptions however, which makes weighting the addresses a bit harder than it should be on the destination service side. First, two source proxies can listen on the same service, because of this, I think it is infeasible to add the weighting logic on subscription. Second, we have to also consider updates -- while on update we can weight a endpoint based on the current endpoints that already exist in a stream, there is a high chance there might be no available endpoints, so inferring weights from existing endpoints when updating also does not sound feasible (even if there is a 1:1 relation between a source and a service stream).
+  * My suggested approach here is to assign weights to endpoints on the proxy side. This could be done by having a one time `GetTopology()` call to the destination service with the service FQDN/IP and the topology token. The destination service can send to the proxy its node's topology labels (based on the sent token), and the service topology preference. This can be kept in memory and calculate weights for endpoints on the fly. The advantage here is that updates will be easy to process and it would ensure backwards compatibility by being slightly isolated from the rest of the discovery api. Endpoints would arrive from the stream would their own topology labels, since those would be set by the watcher.
+  
+* **Destination Service Server**: make changes to the api bindings to accept new topology call from the proxy side.
+
+
+
+### Constraints
+[constraints]: #constraints
+---
+
+* Both `service topology` and `endpoint slices` are not active in a cluster by default. Since they are alpha (and beta) features respectively, they need to be activated by a feature gate. Earlier in the design section I mentioned that we shouldn't use an endpoint watcher and an endpoint slices watcher at the same time. My suggestion is to include a feature flag in the control plane config map. We can access the config map when we create a new server, based on this, when we fire up the watchers we can choose based on the flag which watcher to create.
```

Comment:
Thank you for the pointer! That's a sensible way to do it and won't introduce any configuration. 

I think an `EndpointSlices` watcher should be used instead of an `Endpoints` watcher if the former is activated in a cluster. However, this approach has a major drawback if the `Endpoints` are created manually for a service (without the use of a pod selector); in this case, the `EndpointSlices` will not be created and it's possible the endpoints for a service will not be registered. My question is: would this edge case be of concern to us and do we need to account for it, or would it be up to the adopter to ensure the `EndpointSlices` are created if the service does not include a pod selector?

Reply:
This is a good question.  I'd be curious to know how kube-proxy handles this situation.  Does it watch both Endpoints and EndpointSlices and how does it reconcile differences between the two?

Reply:
I've made all recommended changes to the RFC. I've made a note to look into this specific case and how the `EndpointSlices` are implemented in the service topology PR, will report back once I know a bit more about it.